File: cpp/bob/bob.cpp
#include "bob.h"

bool is_question(std::string remark) {
	return (remark.back() == '?');
}

bool is_yell(std::string remark) {
	bool flag = false;
	for (int i = 0; i < remark.length(); i++) {
		if (isalpha(remark[i])) {
			if (isupper(remark[i])) {
				flag = true;
			} else {
				return false;
			}
		}
	}
	return flag;
}

std::string remove_whitespace(std::string remark) {
	remark.erase(remove_if(remark.begin(), remark.end(), isspace), remark.end());
	return remark;
}

std::string bob::hey(std::string remark) {
	remark = remove_whitespace(remark);
	std::string response = "Whatever.";

	if (remark.empty()) {
		response = "Fine. Be that way!";
	} else if (is_yell(remark)) {
		response = "Whoa, chill out!";
	} else if (is_question(remark)) {
		response = "Sure.";
	}

	return response;
}

File: ruby/bob/bob.rb
class Bob

  class << self 
    def hey(remark)
      if  !/\w+/.match(remark)  
        'Fine. Be that way!'
      elsif remark == remark.upcase && remark[/[a-zA-Z]/]
        'Whoa, chill out!'
      elsif remark[-1] == '?'
        'Sure.'
      else
        'Whatever.'
      end
    end
  end
end